providers/asrockrack/firmware.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	// 2. set update parameters to preserve configurations
+	a.log.V(2).WithValues("step", "2/3").Info("set BIOS preserve flash configuration")
+	err = a.biosUpgradeConfiguration(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed in step 2/3 - set flash configuration")
+	}
+
+	// 3. run upgrade
+	a.log.V(2).WithValues("step", "3/3").Info("proceed with BIOS firmware install")
+	err = a.upgradeBIOS(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed in step 3/3 - proceed with BIOS firmware install")
+	}
+
+	return nil
+}
+
+// firmwareUpdateBIOSStatus returns the BIOS firmware install status
+func (a *ASRockRack) firmwareUpdateStatus(ctx context.Context, component string, installVersion string) (status string, err error) {
+	var endpoint string
+	switch component {
+	case devices.SlugBIOS:
+		endpoint = "api/asrr/maintenance/BIOS/flash-progress"
+	case devices.SlugBMC:
+		endpoint = "api/maintenance/firmware/flash-progress"
+	default:
+		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstallStatus, "component unsupported: "+component)
+	}
+
+	// 1. query the flash progress endpoint
+	//
+	// once an update completes/fails this endpoint will return 500
+	progress, err := a.flashProgress(ctx, endpoint)
+	if err != nil {
+		a.log.V(3).Error(err, "bmc query for install progress returned error: "+err.Error())
+	}
+
+	if progress != nil {
+		switch progress.State {
+		case 0:
+			return devices.FirmwareInstallRunning, nil
+		case 2:
+			return devices.FirmwareInstallComplete, nil
+		default:
+			a.log.V(3).WithValues("state", progress.State).Info("warn", "bmc returned unknown flash progress state")
+		}
+	}
+
+	// 2. query the firmware info endpoint to determine the update status
+	//
+	// at this point the flash-progress endpoint isn't returning useful information
+	var installStatus int
+
+	installStatus, err = a.versionInstalled(ctx, component, installVersion)
+	if err != nil {
+		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstallStatus, err.Error())
+	}
+
+	switch installStatus {
+	case versionStrMatch:
+		return devices.FirmwareInstallComplete, nil
+	case versionStrEmpty:
+		return devices.FirmwareInstallUnknown, nil
+	case versionStrMismatch:
+		return devices.FirmwareInstallRunning, nil
+	}
+
+	return devices.FirmwareInstallUnknown, nil
+}
+
+// versionInstalled returns int values on the status of the firmware version install
+//
+// - 0 indicates the given version parameter matches the version installed
+// - 1 indicates the given version parameter does not match the version installed
+// - 2 the version parameter returned from the BMC is empty (which means the BMC needs a reset)
+func (a *ASRockRack) versionInstalled(ctx context.Context, component, version string) (status int, err error) {
+	fwInfo, err := a.firmwareInfo(ctx)
+	if err != nil {
+		err = errors.Wrap(err, "error querying for firmware info: ")
+		a.log.V(3).Info("warn", err.Error())
+		return versionStrError, err
+	}
+
+	var installed string
+
+	switch component {
```

Comment:
Can you do the validation before the network call?

Reply:
good call, fixed in 0f7a4c72